AEP/aep-00007.md

```diff
@@ -0,0 +1,42 @@
+# AEP 7: Queue system interface design
+
+|                     |ArmoniK Enhancement Proposal|
+|---:                 |:--- |
+| **AEP**             | 7 |
+| **Title**           | Queue system interface design |
+| **Author**          | Jérôme Gurhem <<jgurhem@aneo.fr>> |
+| **Status**          | Active |
+| **Type**            | Standard |
+| **Creation Date**   | 2025-01-15 |
+
+# Abstract
+
+This proposal outlines the rationale and design decisions behind the current message processing and lifecycle management system in ArmoniK. It highlights the primary reasons for adopting a pull-based model, discusses design constraints, and evaluates the trade-offs involved in ensuring reliability and robustness for long-running tasks.
+
+# Motivation
+
+The main motivation for the current architecture is to maintain a historical interface that provides comprehensive traceability and monitoring of tasks and message processing. This interface is central to operational observability and is a key requirement for the system.
+
+# Rationale
+
+## Strict Control of Message Processing
+
+The system is designed to enforce strict control over the number of messages being processed and their lifecycle. Each message in the system represents an executable task with a well-defined processing flow. An executable task is a task whose dependencies were completed, thus being queued for execution. When a message is received, the corresponding task begins execution. The message is acknowledged only after the task completes, regardless of whether it succeeds or fails. In cases where the queueing service loses its connection with the agent before the message is acknowledged, the system ensures reliability by redelivering the message. This guarantees that tasks are not lost and are executed even when errors or interruptions occur.
+
+Retries for tasks are handled by treating them as new tasks. Each retry is introduced as a new message, which simplifies tracking and execution. Messages are acknowledged when the task completes successfully and subtasks are submitted. ArmoniK processes new messages only after the current task has been dispatched to a worker, starting the acquisition of a new task during the processing of the previous one. This design allows to have a task ready for execution as soon as the previous task ends.
```

Comment:
I'd rewrite  "Messages for retry tasks are acknowledged when ..." just to be more explicit that in this case we do care about task completion success.  And the reference to subtasking confuses me, since is not at all mentioned on the previous paragraph addressing "normal tasks".